diff --git a/src/WebformQuery.php b/src/WebformQuery.php
index 29099e7..b14bb81 100644
--- a/src/WebformQuery.php
+++ b/src/WebformQuery.php
@@ -45,7 +45,7 @@ class WebformQuery {
    */
   public function setWebform($webform_id = NULL) {
     if (!is_null($webform_id)) {
-      $this->addCondition('webform_id', $webform_id);
+      $this->addCondition('webform_id', $webform_id, '=', 'webform_submission');
     }
     return $this;
   }
@@ -59,10 +59,16 @@ class WebformQuery {
    *   Value to compare.
    * @param mixed $operator
    *   Operator.
+   * @param string $table
+   *   The table to query.
    *
    * @return $this
    */
-  public function addCondition($field, $value = NULL, $operator = '=') {
+  public function addCondition($field, $value = NULL, $operator = '=', $table = 'webform_submission_data') {
+    // Clean up table and field as placeholders can't be used for them.
+    $table = $this->connection->escapeTable($table);
+    $field = $this->connection->escapeTable($field);
+
     // Check for webform_id.
     if ($field === 'webform_id') {
       // Check for existing condition at 0.
@@ -73,6 +79,7 @@ class WebformQuery {
         'field' => $field,
         'value' => $value,
         'operator' => $operator,
+        'table' => $table,
       ];
     }
     else {
@@ -90,6 +97,7 @@ class WebformQuery {
           'field' => $field,
           'value' => $value,
           'operator' => $operator,
+          'table' => $table,
         ];
       }
     }
@@ -125,6 +133,17 @@ class WebformQuery {
    *   Array of objects with one property: sid
    */
   public function execute() {
+    // Return the results.
+    return $this->processQuery()->fetchAll();
+  }
+
+  /**
+   * Process the query and return a database statement.
+   *
+   * @return Drupal\Core\Database\Statement
+   *   Prepared statement.
+   */
+  public function processQuery() {
     // Generate query elements from the conditions.
     $query_elements = $this->buildQuery();
 
@@ -133,38 +152,29 @@ class WebformQuery {
     $this->sort = [];
 
     // Execute the query.
-    $response = $this->connection->query($query_elements['query'], $query_elements['values']);
-
-    // Return the results.
-    return $response->fetchAll();
+    return $this->connection->query($query_elements['query'], $query_elements['values']);
   }
 
   /**
    * Build the query from the conditions.
    */
   public function buildQuery() {
-    $query = 'SELECT DISTINCT sid FROM {webform_submission_data} wsd';
-    $values = [];
+    $query = '';
+
     foreach ($this->conditions as $key => $condition) {
-      // Check if it's the first condition.
-      if ($key === 0) {
-        // Check for database field webform_id.
-        if ($condition['field'] == 'webform_id') {
-          $query .= ' WHERE wsd.webform_id ' . $condition['operator'] . ' :' . $condition['field'];
-        }
-        else {
-          $query .= ' WHERE wsd.name = :' . $condition['field'] . '_name AND wsd.value ' . $condition['operator'] . ' :' . $condition['field'];
-          $values[':' . $condition['field'] . '_name'] = $condition['field'];
-        }
+      $alias = 'table' . $key;
+
+      // Check if querying the results table.
+      if ($condition['table'] !== 'webform_submission_data') {
+        $query = ' AND sid IN (SELECT sid FROM {' . $condition['table'] . '} ' . $alias . ' WHERE ' . $alias . '.' . $condition['field'] . ' ' . $condition['operator'] . ' :' . $condition['field'] . $key . ')' . $query;
       }
       else {
         // Normal condition for a webform submission field.
-        $alias = 'wsd' . $key;
-        $query .= ' AND sid IN (SELECT sid from {webform_submission_data} ' . $alias . ' WHERE ' . $alias . '.name = :' . $condition['field'] . '_name';
-        $query .= ' AND ' . $alias . '.value ' . $condition['operator'] . ' :' . $condition['field'] . ')';
-        $values[':' . $condition['field'] . '_name'] = $condition['field'];
+        $query .= ' AND sid IN (SELECT sid FROM {webform_submission_data} ' . $alias . ' WHERE ' . $alias . '.name = :' . $condition['field'] . '_name' . $key;
+        $query .= ' AND ' . $alias . '.value ' . $condition['operator'] . ' :' . $condition['field'] . $key . ')';
+        $values[':' . $condition['field'] . '_name' . $key] = $condition['field'];
       }
-      $values[':' . $condition['field']] = $condition['value'];
+      $values[':' . $condition['field'] . $key] = $condition['value'];
     }
 
     // Check for sort criteria.
@@ -179,6 +189,9 @@ class WebformQuery {
       $query .= ' ORDER BY (SELECT ' . $orderby_alias . '.value FROM {webform_submission_data} ' . $orderby_alias . ' WHERE ' . $orderby_alias . '.name=\'' . $orderby['field'] . '\' AND ' . $orderby_alias . '.sid=wsd.sid) ' . $orderby['direction'];
     }
 
+    $query = substr_replace($query, ' WHERE', 0, 4);
+    $query = 'SELECT sid FROM {webform_submission} ws' . $query;
+
     return ['query' => $query, 'values' => $values];
 
   }
